test/libsolidity/syntaxTests/functionTypes/declaration_type_conversion.sol

```diff
@@ -0,0 +1,11 @@
+contract D {
+	function f() external {}
+	function g() external {}
+}
+contract C {
+	function f(bool c) public pure {
+		(c ? D.f : D.g);
```

Comment:
What about:
```solidity
library L {
    function f() internal pure {}
    function g() internal pure {}
}

contract C {
    function f(bool b) public pure {
        (b ? L.f : L.g);
    }
}
```
Shouldn't this also be disallowed?

Reply:
Why? Is there a problem with it? These are perfectly fine internal function pointers, I'd think that works just fine and is perfectly valid.

Reply:
Ah right :)

Reply:
I'm actually wondering what other cases there are where we might get `Kind::Declaration`. I see in the code that `Kind::Declaration` is the default and never assigned explicitly, which makes it a little hard to track down.

From user's prespective, what's a "function declaration type"? Is that even well defined?

Reply:
Yeah, I have similar questions as well. Testing locally I could see that both cases below also works:
```solidity
contract D {
    function g() external {}
}

contract C {
    struct S {
        function() external h;
    }

    S private s;

    function f(bool b) public pure {
        (b ? s.h : D.g);
    }
}
// ----
// TypeError 1080: (175-188): True expression's type function () external does not match false expression's type function D.g().
```
And:
```solidity
interface I {
    function f() external;
    function g() external;
}

contract C {
    function f(bool b) public pure {
        (b ? I.f : I.g);
    }
}
// ----
// TypeError 1080: (130-143): True expression's type function I.f() does not match false expression's type function I.g().

```

Reply:
``Kind::Declaration`` is a function type that doesn't refer to an actual function pointer, but merely to a function declaration.
That happens just in case you refer to a function with a member access to a container type, not to an object of the container type.

So it's basically two cases:
- Accesses to types that contain external functions visible via the outer declaration:
  ```solidity
  contract C {
    function f() external {}
  }
  function g() {
    C.f; // FunctionKind::Declaration
  }
  ```
- References to internal base functions by base contract name - I **think** at least :-D:
  ```solidity
  contract C {
    function f() internal {}
  }
  contract D is C {
    function g() {
      C.f; // FunctionKind::Declaration
    }
  }
  ```

For the latter case, if it turns out that they're actually ``FunctionKind::Declaration``, we need to check if the fix actually works as intended. For those we actually *may* want to keep them convertible, but it may not be easy to distinguish.

Reply:
> For the latter case, if it turns out that they're actually FunctionKind::Declaration

I added a test case to confirm that they are of `Kind::Internal` and **not** of `Kind::Declaration`.

Reply:
> What about @r0qs's case but with functions marked external? IIRC you cannot take pointers to external library functions.

To answer my own question: that's `Kind::Delegatecall`.

> - References to internal base functions by base contract name - I **think** at least :-D: 

I just checked and that one's actually `Kind::Internal` according to AST:

```json
"typeIdentifier": "t_function_internal_nonpayable$__$returns$__$"
```